### PR TITLE
Create a table-based result data structure

### PR DESCRIPTION
**Pull request by jeremykubica**
The goal is to address #535. The `Results` object provides a wrapper around an astropy table that lets the user easily build it from a list of trajectories and to update multiple columns at once. It preserves the tracking (and ability to revert) filtering of results.

A user can directly access the table with:
```
my_table = table.results
```
or access the columns as:
``table["x"]``
directly interact with it, filter it, add columns, etc. without having to later sync it back.

### CODE REVIEW

**Comment by jeremykubica** [comment]
If we don't care about tracking the filtered rows and providing the ability to revert those, we could simplify this a bunch further.

File: tests/test_result_table.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        self.num_entries = 10
+        self.trj_list = [
+            make_trajectory(x=i, y=i + 0, vx=i - 2.0, vy=i + 5.0, flux=5.0 * i, lh=100.0 + i, obs_count=i)
+            for i in range(self.num_entries)
+        ]
+
+    def test_create(self):
+        table = ResultTable(self.trj_list)
+        self.assertEqual(len(table), self.num_entries)
+        for i in range(self.num_entries):
```
Comment (DinoBektesevic) [inline tests/test_result_table.py:25]: I feel like maybe dumping something into `utils.py` in the tests that just mocks out a few of these results could be helpful to remove a lot of this repeated code (including comparison code). But I have to go teach so I hadnt paid that much attention to tests.

**Review by DinoBektesevic — COMMENTED** [review]
I think this is ok, obviously it changes one pretty big thing - jagged array support is removed because of the fundamental change to how the data is represented internally. 

Mostly my wants were related to `__getitem__` and `__repr_html_` because that's what I end up looking at most often. 

This change did bring up a lot of questions for me about the utility and purpose of `Trajectory` in **this** particular context. I think there's a lot of structure of arrays versus array of structures conflicts in this proof of concept that I'd like to see cleared out by making a definitive decision to go one or the other way - this double dipping causes some indirection that are hard to follow. 

I left a bunch of little comments for documentation but it's not serious, please feel free to ignore I understand this is a POC implementation.

File: src/kbmod/result_table.py [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+import numpy as np
+from pathlib import Path
+
+from astropy.table import Table, vstack
+
+from kbmod.trajectory_utils import make_trajectory, update_trajectory_from_psi_phi
+from kbmod.search import Trajectory
+
+
+class ResultTable:
```
[the remark this answers is not in the document]
Comment (jeremykubica) [reply src/kbmod/result_table.py:14]: Sure. That's not a problem.

File: src/kbmod/filters/clustering_filters.py [hunk 3]
```diff
@@ -109,8 +113,14 @@ def _build_clustering_data(self, result_list):
...
-        x_arr = np.array(result_list.get_result_values("trajectory.x")) / self.width
-        y_arr = np.array(result_list.get_result_values("trajectory.y")) / self.height
+        if type(result_data) is ResultList:
+            x_arr = np.array(result_data.get_result_values("trajectory.x")) / self.width
+            y_arr = np.array(result_data.get_result_values("trajectory.y")) / self.height
+        elif type(result_data) is Results:
+            x_arr = np.array(result_data["x"]) / self.width
+            y_arr = np.array(result_data["y"]) / self.height
+        else:
+            raise TypeError("Unknown data type for clustering.")
```
[the remark this answers is not in the document]
Comment (jeremykubica) [reply src/kbmod/filters/clustering_filters.py:123]: Correct. Want to get this fully integrated and then remove all instances of `ResultList`. This is just temporary so I can break it into a few PRs.

File: src/kbmod/results.py [hunk 4]
```diff
@@ -0,0 +1,591 @@
...
+            if trj.valid:
+                input_d["x"].append(trj.x)
+                input_d["y"].append(trj.y)
+                input_d["vx"].append(trj.vx)
+                input_d["vy"].append(trj.vy)
+                input_d["likelihood"].append(trj.lh)
+                input_d["flux"].append(trj.flux)
+                input_d["obs_count"].append(trj.obs_count)
+                num_valid += 1
+            elif track_filtered:
```
[the remark this answers is not in the document]
Comment (jeremykubica) [reply src/kbmod/results.py:130]: It is add the valid trajectories to `input_d` whenever it encounters one. It is only adding the invalid trajectories to `invalid_d` if it is going to use track the items it filtered out.

If `track_filtered` is `True`: Then `input_d` is all the valid trajectories and `invalid_d` is all the invalid ones.
Otherwise (when `track_filtered==False`): Then `input_d` is all the valid trajectories and `invalid_d` is empty.

File: src/kbmod/results.py [hunk 5]
```diff
@@ -0,0 +1,591 @@
...
+        psi = self.table["psi_curve"]
+        phi = self.table["phi_curve"]
+
+        # Create a mask of valid data.
+        valid = (phi != 0) & np.isfinite(psi) & np.isfinite(phi)
+        if filter_indices and "index_valid" in self.table.colnames:
+            valid = valid & self.table["index_valid"]
+
+        lh_matrix = np.full(psi.shape, mask_value)
+        lh_matrix[valid] = np.divide(psi[valid], np.sqrt(phi[valid]))
```
[the remark this answers is not in the document]
Comment (jeremykubica) [reply src/kbmod/results.py:265]: `/` should work. I will change it.

File: src/kbmod/results.py [hunk 6]
```diff
@@ -0,0 +1,591 @@
...
+    # (column name, dype, default value) for each required column.
+    _required_cols = [
+        ("x", "int64", 0),
+        ("y", "int64", 0),
+        ("vx", "float64", 0.0),
+        ("vy", "float64", 0.0),
+        ("likelihood", "float64", 0.0),
+        ("flux", "float64", 0.0),
+        ("obs_count", "int64", 0),
+    ]
```
[the remark this answers is not in the document]
Comment (jeremykubica) [reply src/kbmod/results.py]: Changed. Although I needed to use `float`, `int`, etc because the numpy versions are deprecated in favor of those.

File: src/kbmod/results.py [hunk 7]
```diff
@@ -0,0 +1,591 @@
...
+        if data is None:
+            # Set up the basic table meta data.
+            self.table = Table(
+                names=[col[0] for col in self._required_cols],
+                dtype=[col[1] for col in self._required_cols],
+            )
+        elif type(data) is dict:
+            self.table = Table(data)
+        elif type(data) is Table:
+            self.table = data.copy()
```
[the remark this answers is not in the document]
Comment (jeremykubica) [reply src/kbmod/results.py]: Done

File: src/kbmod/results.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (DinoBektesevic) [reply src/kbmod/results.py:130]: Ah, I see. ok, no problems.

File: src/kbmod/trajectory_utils.py [hunk 8]
```diff
@@ -46,13 +46,13 @@ def make_trajectory(x=0, y=0, vx=0.0, vy=0.0, flux=0.0, lh=0.0, obs_count=0):
...
-    trj.x = x
-    trj.y = y
+    trj.x = int(x)
+    trj.y = int(y)
     trj.vx = vx
     trj.vy = vy
     trj.flux = flux
     trj.lh = lh
-    trj.obs_count = obs_count
+    trj.obs_count = int(obs_count)
```
[the remark this answers is not in the document]
Comment (DinoBektesevic) [reply src/kbmod/trajectory_utils.py]: What was the idea here if marked resolved?

File: src/kbmod/results.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (DinoBektesevic) [reply src/kbmod/results.py]: Interesting, the recommendation is to insist on using `np.float64` and `np.float32` instead of system default alias `np.float`. Good to know, thanks:

https://numpy.org/devdocs/release/1.20.0-notes.html#using-the-aliases-of-builtin-types-like-np-int-is-deprecated

**Review by DinoBektesevic — APPROVED** [review]
Looks amazing! Thanks for all the effort.
